src/limb.rs

```diff
@@ -0,0 +1,298 @@
+    // and switch to the other mechanism.
+    //
+    // TODO This interface is a bit unfortunate, since it requires the caller to
+    // ensure the destination byte array is the right size. But different
+    // callers need the result in byte arrays of different fixed size, and AFAIK
+    // we can't support that better without language changes. See:
+    // https://github.com/rust-lang/rfcs/issues/1038
+    //
+    /// Chooses a positive integer within the range and stores it into `dest`
+    /// encoded as big-endian bytes.
+    ///
+    /// This function is intended to be suitable for generating private keys.
+    pub fn sample_into_bytes(&self, dest: &mut [u8], rng: &rand::SecureRandom)
+                             -> Result<(), ()> {
+        assert!(self.valid().is_ok());
+
+        // NSA Guide Steps 1, 2, and 3.
+        let num_bytes = self.bytes_count();
+        assert!(dest.len() >= num_bytes);
+
+        let mut limbs: [Limb; MAX_RANGE_LIMBS] = [0; MAX_RANGE_LIMBS];
+
+        // XXX: The value 100 was chosen to match OpenSSL due to uncertainty of
+        // what specific value would be better, but it seems bad to try 100 times.
+        for _ in 0..100 {
+
+            // NSA Guide Step 4.
+            //
+            // The requirement that the random number generator has the requested
+            // security strength is delegated to `rng`.
+            try!(rng.fill(&mut dest[..num_bytes]));
+            dest[0] &= 0xff >> (self.leading_zero_bits() % 8);
+
+            // NSA Guide Steps 5, 6, and 7.
+            //
+            // XXX: The NSA guide says that we should verify that the random scalar
+            // is in the range [0, n - 1) and then add one to it so that it is in
+            // the range [1, n). Instead, we verify that the scalar is in the range
+            // [1, n) like BoringSSL (et al.) does. This way, we avoid needing to
+            // compute or store the value (n - 1), we avoid the need to implement
+            // a function to add one to a scalar, and we avoid needing to convert
+            // the scalar back into an array of bytes. TODO: Is there any security
+            // advantage to the way the NSA suggests? There doesn't seem to be,
+            // other than being less error prone w.r.t. accidentally generating
+            // zero-valued keys.
+            set_limbs_from_big_endian_bytes(&mut limbs[..self.limbs.len()],
+                                        &dest[..num_bytes]);
+            if !self.are_limbs_within(&limbs[..self.limbs.len()]) {
+                continue;
+            }
+
+            // NSA Guide Step 8 is done in `private_to_public()`.
+
+            // NSA Guide Step 9.
+            return Ok(());
+        }
+
+        Err(())
+    }
+
+}
+
+
+/// Initializes little-endian limbs from big-endian bytes.
+///
+/// Equivalent to `BN_bin2bn`.
+pub fn set_limbs_from_big_endian_bytes(limbs: &mut[Limb], bytes: &[u8]) {
```

Comment:
> parse_big_endian_value. I think we should unify these together. 

Actually, I guess the point of this function to to be more clearly constant time, which is presumably why there were two functions to start with. So, it is OK if we don't unify them together. Again, see my other comment; I'm kind of hoping we can just avoid this function completely.